lib/tables.c

```diff
@@ -4675,3 +4675,146 @@ table_collection_simplify(table_collection_t *self,
     simplifier_free(&simplifier);
     return ret;
 }
+
+
+/*******************
+ * Cleaning tables *
+ ******************/
+
+typedef struct {
+    site_table_t *sites;
+    mutation_table_t *mutations;
+    /* Map of old site IDs to new site IDs. */
+    site_id_t *site_id_map;
+} table_cleaner_t;
+
+static int table_cleaner_run(table_cleaner_t *self)
+{
+    // Remove any sites with duplicate positions, retaining only the *first*
+    // one. Assumes the tables have been sorted, throwing an error if not.
+    // TODO: this could be done in place rather than by constructing a new
+    // table.
+
+    int ret = 0;
+
+    if (self->sites->num_rows <= 1) {
+        // nothing to do
+        goto out;
+    }
+
+    table_size_t j, k, site_j;
+    table_size_t as_length, as_offset;
+    table_size_t md_length, md_offset;
+    double last_position, position;
+
+    self->site_id_map[0] = 0;
+    site_j = 0;
+    last_position = self->sites->position[0];
+    as_offset = self->sites->ancestral_state_offset[1];
+    md_offset = self->sites->metadata_offset[1];
+    for (j = 1; j < self->sites->num_rows; j++) {
+        position = self->sites->position[j];
+        if (position < last_position) {
+            ret = MSP_ERR_UNSORTED_SITES;
+            goto out;
+        }
+        if (position != last_position) {
+            site_j++;
+            as_length = self->sites->ancestral_state_offset[j + 1] - self->sites->ancestral_state_offset[j];
+            md_length = self->sites->metadata_offset[j + 1] - self->sites->metadata_offset[j];
+            if (site_j != j) {
+                assert(site_j < j);
+                self->sites->position[site_j] = self->sites->position[j];
+                memcpy(self->sites->ancestral_state + self->sites->ancestral_state_offset[site_j],
+                        self->sites->ancestral_state + self->sites->ancestral_state_offset[j],
+                        as_length);
+                self->sites->ancestral_state_offset[site_j] = as_offset;
+                memcpy(self->sites->metadata + self->sites->metadata_offset[site_j],
+                        self->sites->metadata + self->sites->metadata_offset[j],
+                        md_length);
+                self->sites->metadata_offset[site_j] = md_offset;
+            }
+            as_offset += as_length;
+            md_offset += md_length;
+            last_position = position;
+        }
+        self->site_id_map[j] = site_j;
+    }
+
+    if (site_j == j) {
```

Comment:
OK, this is definitely ``goto`` as control flow! Change to
```c
if (site_j != j) { 
    // reset site table size
    self->sites->num_rows = site_j + 1;
```
Or, you could just leave it fall through? The fewer paths through the code the better IMO, I don't really mind if a it's a little less efficient in some corner cases because of this. (However, simplicity is more important --- if having fewer paths through the function makes the code more complicated, then I'd go with the simpler and easier to understand option.)

Reply:
not sure what you mean by "fall through" -- here we are avoiding looping through every entry in the mutation table and setting its' `site` equal to itself -- if the tables are big, this seems not insignificant, and easy to avoid?

Reply:
Sure, you're right. I was just thinking in a big-O sense there's no real difference between iterating over the tables once and twice. But I think this it's more readable the way you have it and it is an easily avoidable cost.